### ISSUE TITLE
Question about Objective from User

### ISSUE THREAD

**Issue by mohawk811** [PlasmaControl/DESC#1735]
```
    def myfun3(grid, data):
        modB = data["|B|"]  # shape: (ntheta, nzeta, nr) or similar
        B_max_r = data['min_tz |B|']  # shape: (nr,)
        B_min_r = data['max_tz |B|'] # shape: (nr,)
        
        mirror_r = (B_max_r - B_min_r) / B_max_r  # shape: (nr,)
        
        return (mirror_r[-1] - mirror_r[0]) / mirror_r[-1]
```

I was wondering If I could some advice on this custom function I am trying to create, where I redefine Mirror ratio. My question is how would I provide upper and lower bounds to max and min of mod B for a given surface. My goal is have the optimizer converge to an equilibrium that has a positive radial(flux coordinates) slope in Bmax and Bmin. Attached is plot from (https://arxiv.org/pdf/2306.17506) that depicts the kind of feature I would like to optimize for.

<img width="266" alt="Image" src="https://github.com/user-attachments/assets/33f90ed9-3dff-4eff-b855-06936122341d" />

**Comment by f0uriest** [comment]
I think you could do something like this:
```python
def myfun3(grid, data):
    B_max = data['min_tz |B|']  # shape: (nr,)
    B_min = data['max_tz |B|'] # shape: (nr,)
    dB_max_dr = jnp.diff(B_max)
    dB_min_dr = jnp.diff(B_min)
    return jnp.concatenate([dBmax_dr, dBmin_dr])
```
and then when creating the objective set `bounds=(0, np.inf)`

Basically you're taking a first order finite difference approximation to the radial slope of B{max,min} and then saying that you want that to be positive everywhere

**Comment by mohawk811** [comment]
The Issue that I am having is that i can easily get a positive slope in Bmax but I'm consistently getting a negative slope in Bmin.

<img width="566" alt="Image" src="https://github.com/user-attachments/assets/ce2484ab-6e26-42d3-aff9-11a47226fee4" />

<img width="374" alt="Image" src="https://github.com/user-attachments/assets/96216fb0-5a51-4231-a72c-a243dfd47f3f" />

**Comment by dpanici** [comment]
Bmin I can see intuitively having difficulty in getting a positve slope with rho, simply due to the inherent `1/R` effect of toroidicity on the magnetic field strength. I think basically any finite aspect ratio toroidal plasma (be it tokamak or stellarator) has a stronger field on the inboard side as opposed to the outboard side. 

As `rho` increases, what you are seeing is that Bmax is able to increase, which make sense as larger rho corresponds to flux surfaces which have areas of smaller `R` (i.e. more inboard then the magnetic axis). 

But at the same time, these flux surfaces also have areas of higher `R` (i.e. more outboard than the magnetic axis), and these areas tend to have smaller magnitude B just due to toroidal effects

I guess my advice would be increasing the weight on the objective, or perhaps starting from an initial guess that is closer to what you are looking for

**Comment by f0uriest** [comment]
For the second one, assuming `r` is some scalar hyperparameter I think you could do it like
```python
def foo(grid, data):
    modB = grid.meshgrid_reshape(data['|B|'], "rtz") # shape (rho, theta, zeta)
    mean_modB = grid.meshgrid_reshape(data['<|B|>'], "rtz")[:,0,0] # shape (rho)
    return (modB / mean_modB[:,None,None]).flatten()
```

and then in the objective set
```python
bounds=[2/(r+1), 2*r/(r+1)]```

**Comment by mohawk811** [comment]
I tried feeding r as 

`r=eq.compute('mirror ratio')['mirror ratio']`

with the bounds as 

`bounds=[2/(r+1), 2*r/(r+1)]`

but I get 
````
File [/opt/anaconda3/envs/desc/lib/python3.12/site-packages/desc/objectives/_generic.py:519](http://localhost:8888/opt/anaconda3/envs/desc/lib/python3.12/site-packages/desc/objectives/_generic.py#line=518), in ObjectiveFromUser.__init__(self, fun, thing, target, bounds, weight, normalize, normalize_target, loss_function, deriv_mode, grid, name, jac_chunk_size, compute_kwargs, **kwargs)
    517 self._grid = grid
    518 self._compute_kwargs = setdefault(compute_kwargs, {})
--> 519 super().__init__(
    520     things=thing,
    521     target=target,
    522     bounds=bounds,
    523     weight=weight,
    524     normalize=normalize,
    525     normalize_target=normalize_target,
    526     loss_function=loss_function,
    527     deriv_mode=deriv_mode,
    528     name=name,
    529     jac_chunk_size=jac_chunk_size,
    530 )
    531 self._p = _parse_parameterization(thing)

File [/opt/anaconda3/envs/desc/lib/python3.12/site-packages/desc/objectives/objective_funs.py:1066](http://localhost:8888/opt/anaconda3/envs/desc/lib/python3.12/site-packages/desc/objectives/objective_funs.py#line=1065), in _Objective.__init__(self, things, target, bounds, weight, normalize, normalize_target, loss_function, deriv_mode, name, jac_chunk_size)
   1064 assert normalize in {True, False}
   1065 assert normalize_target in {True, False}
-> 1066 assert (bounds is None) or (isinstance(bounds, tuple) and len(bounds) == 2)
   1067 assert (bounds is None) or (target is None), "Cannot use both bounds and target"
   1068 assert loss_function in [None, "mean", "min", "max"]

AssertionError:
````

**Comment by dpanici** [comment]
- Bounds has to be a tuple, which explains the assertion error
- make sure you know what you are putting here, just fyi: `r=eq.compute('mirror ratio')['mirror ratio']` is itself an array consisting of the mirror ratio per surface on whatever the default radial resolution grid we compute is (probably `LinearGrid(L=eq.L_grid)`).  As it stands you are giving a bounds that is some array value, I suspect you want to instead pass it in for either a single surface or some set of surfaces you've chosen beforehand (so you should pass a grid into eq.compute is what I am saying)

**Comment by YigitElma** [comment]
I will close the issue due to inactivity, but feel free to re-open it if you have more questions!
